lib/aba/batch.rb

```diff
@@ -47,56 +47,64 @@ def initialize(attrs = {}, transactions = [])
     end
 
     def to_s
-      raise RuntimeError, 'No transactions present - add one using `add_transaction`' if @transactions.empty?
+      raise RuntimeError, 'No entries present - add one using `add_transaction` or `add_return`' if entries.empty?
       raise RuntimeError, 'ABA data is invalid - check the contents of `errors`' unless valid?
 
       # Descriptive record
       output = "#{descriptive_record}\r\n"
 
       # Transactions records
-      output += @transactions.map{ |t| t[1].to_s }.join("\r\n")
+      output += entries.map { |t| t[1].to_s }.join("\r\n")
 
       # Batch control record
       output += "\r\n#{batch_control_record}"
     end
 
     def add_transaction(attrs = {})
-      if attrs.instance_of?(Aba::Transaction)
-        transaction = attrs
-      else
-        transaction = Aba::Transaction.new(attrs)
-      end
-      @transactions[@transaction_index] = transaction
-      @transaction_index += 1
-      transaction
+      add_entry(Aba::Transaction, attrs)
+    end
+
+    def add_return(attrs = {})
+      add_entry(Aba::Return, attrs)
+    end
+
+    def transactions
+      entries.select { |entry| entry.instance_of?(Aba::Transaction) }
     end
 
     def transactions_valid?
-      !has_transaction_errors?
+      !transactions.map { |t| t[1].valid? }.include?(false)
     end
 
     def valid?
-      !has_errors? && transactions_valid?
+      !has_errors? && !has_entry_errors?
     end
 
     def errors
       # Run validations
       has_errors?
-      has_transaction_errors?
+      has_entry_errors?
 
       # Build errors
       all_errors = {}
       all_errors[:aba] = self.error_collection unless self.error_collection.empty?
-      transaction_error_collection = @transactions.each_with_index.map{ |(k, t), i| [k, t.error_collection] }.reject{ |e| e[1].nil? || e[1].empty? }.to_h
-      all_errors[:transactions] = transaction_error_collection unless transaction_error_collection.empty?
+      entry_error_collection = entries.map { |k, t| [k, t.error_collection] }.reject { |e| e[1].nil? || e[1].empty? }.to_h
+      all_errors[:entries] = entry_error_collection unless entry_error_collection.empty?
 
       all_errors unless all_errors.empty?
     end
 
     private
 
-    def has_transaction_errors?
-      @transactions.map{ |t| t[1].valid? }.include?(false)
+    def add_entry(type, attrs)
+      (attrs.instance_of?(type) ? attrs : type.new(attrs)).tap do |entry|
+        entries[@entries_index] = entry
+        @entries_index += 1
+      end
+    end
+
+    def has_entry_errors?
+      entries.map { |t| t[1].valid? }.include?(false)
```

Comment:
Again could we just use `all?`

Reply:
I originally thought the same but then realised there was a subtle difference then confused myself then remembered what the difference was again :P

The problem is that the validation is collecting errors for all fields.
`all?` will return `false` the moment the block returns `false`. That means that if there are multiple errors we'll only an error report for the first one.